### PR TITLE
Compare tractograms (TODI, TDI, ACC)

### PR DESCRIPTION
**Pull request by frheault**
# Quick description

Scripts to compute differences between tractograms (or bundles)
2 big tractograms can be multi processed (2x2GB = 2h at 20 CPU)

It computes TODI+ACC, Correlation of TDI, voxel-wise distances of streamlines (using FSS)

4 outputs
out_acc.nii.gz: How does local orientations compare?
out_correlation.nii.gz How does the density of streamlines agrees?
out_diff.nii.gz How different is the shape of the streamlines?
out_heatmap.nii.gz Power means of 3 others?

## Type of change

Check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

image

# Checklist

- [x] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ ] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:1064]: logging.info('Computing TODI on tractogram 1...')

(idem below for 2)

(You can kepp the # sign but I don't how to stop github to cite the 1st and 2nd PR when I do!!)

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:1076]: Eventually, we should have a method 'compute_todi' that would also be used in scil_tractogram_compute_todi. If you don't do it here, let's add a todo in our google doc or on github.

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:881]: @arnaudbore How do you want to process? Should we describe every global variable?

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:893]: Is it a good choice to use -1 when the correlation does not exist? A correlation of -1 is a number that exists (perfect anti-correlation).

Should you explain in the script or somewhere the possible cases why a correlation would be -1?

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:902]: Again, maybe describe in the script or somewhere all cases why distance could be -1.

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:900]: Your comment says 2mm, your code 1.5, the other comment below, 1. Should it be a variable? Even an option to user, or depending on the resolution of the data?

File: scilpy/tractograms/tractogram_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/tractogram_operations.py:915]: I was hoping to read the code and understand what 'distance' means to you. Well... Can't understand this! Can you add a few comments for us mere mortals?

File: scilpy/tractograms/tractogram_operations.py
Comment (frheault) [reply scilpy/tractograms/tractogram_operations.py:893]: I can also put a 0? I would rather avoid NaN as they often break everything afterward (other scipts).

Do you think 0 is better?

Comment (EmmaRenauld) [reply scilpy/tractograms/tractogram_operations.py:893]: I would think so... But it depends on the use case I guess. Or like, -2, and people know that it's not possible? In future scripts, do you have options to reject some values?

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (frheault) [reply scripts/scil_tractogram_pairwise_comparison.py:30]: Also it is important to note that from big script it is possible that when we extract the main() and put it somewhere, it fits in two places (some script deal with maps, labels and tractogram and it can be difficult to pick the best one)

If you create scilpy.tractograms.reproducibility_measures
Do you create scilpy.bundles.reproducibility_measures ?

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (EmmaRenauld) [inline scripts/scil_tractogram_pairwise_comparison.py:10]: agreement :)

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (EmmaRenauld) [inline scripts/scil_tractogram_pairwise_comparison.py:15]: look

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (frheault) [reply scripts/scil_tractogram_pairwise_comparison.py:95]: --out_dir will force an empty folder, the user could use --out_prefix to put it in a folder next to other files.

I would say, 'it is not a bug, it is a feature'

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (frheault) [reply scripts/scil_tractogram_pairwise_comparison.py:102]: I chose '' as a default since this does not check the current directory if it is empty, it simply creates the files (most common use case). If I put './' and the user uses -f it would delete everything which is dangerous.

Using '' as default is the same as not using --out_dir (write in the current directory (even if not empty), like most scripts).

File: scilpy/image/volume_operations.py
Comment (arnaudbore) [inline scilpy/image/volume_operations.py:466]: add unit test template for these new functions

File: scilpy/tractograms/streamline_operations.py
Comment (arnaudbore) [inline scilpy/tractograms/streamline_operations.py:394]: same here

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (arnaudbore) [reply scripts/scil_tractogram_pairwise_comparison.py:41]: I like scilpy.tractanalysis.reproducibility_measures it seems to fit our needs here. We'll see in the future if we need to create scilpy.tractograms.reproducibility_measures or scilpy.bundles.reproducibility_measures

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (arnaudbore) [inline scripts/scil_tractogram_pairwise_comparison.py:81]: Change this with: 
`logging.getLogger().setLevel(logging.getLevelName(args.verbose))`

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (arnaudbore) [reply scripts/scil_tractogram_pairwise_comparison.py:67]: @frheault a quick sentence here please.

File: scripts/scil_tractogram_pairwise_comparison.py
Comment (arnaudbore) [reply scripts/scil_tractogram_pairwise_comparison.py:95]: Good enough
image

File: scilpy/tractograms/tractogram_operations.py
Comment (arnaudbore) [reply scilpy/tractograms/tractogram_operations.py:883]: We keep it as it is.

**Review by arnaudbore — CHANGES_REQUESTED** [review]
A couple of things to fix but otherwise GTG

File: scilpy/tractograms/tests/test_tractogram_operations.py
Comment (arnaudbore) [reply scilpy/tractograms/tests/test_tractogram_operations.py:18]: I think I've got it, it is for scilpy/tractanalysis/tests/test_reproducibility_measures.py but you can remove it for now.

**Review by arnaudbore — APPROVED** [review]
A quick comment and a pep8 fix then LGTM @EmmaRenauld or @karanphil if you have some time I would need a second review of this code. Thank you in advance

File: scilpy/image/volume_operations.py
Comment (EmmaRenauld) [inline scilpy/image/volume_operations.py:537]: This could actually work with an array of any dimension. It does not have to be a volume. Ex, it is probably useful everytime we want to change the range of a colormap.  Should we put it in utils? @arnaudbore

File: scilpy/tractanalysis/reproducibility_measures.py
Comment (EmmaRenauld) [inline scilpy/tractanalysis/reproducibility_measures.py:131]: Is this an error?

File: scilpy/tractanalysis/reproducibility_measures.py
Comment (EmmaRenauld) [inline scilpy/tractanalysis/reproducibility_measures.py:457]: for the first sft

File: scilpy/tractanalysis/reproducibility_measures.py
Comment (EmmaRenauld) [inline scilpy/tractanalysis/reproducibility_measures.py:465]: for the second sft

File: scilpy/tractanalysis/reproducibility_measures.py
Comment (EmmaRenauld) [inline scilpy/tractanalysis/reproducibility_measures.py:516]: We don't usually put it like this. By default, python always returns tuples, no need to say it. You can remove this line and unindent the two below.

File: scilpy/tractograms/streamline_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/streamline_operations.py:412]: With the idea of not using the hidden _data, could you use:
total_points = np.sum([len(s) for s in sft.streamlines])

File: scilpy/tractograms/streamline_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/streamline_operations.py:413]: Is it dangerous to use the hidden offsets?

**Review by EmmaRenauld — COMMENTED** [review]
Thanks!

File: scilpy/tractograms/streamline_operations.py
Comment (EmmaRenauld) [inline scilpy/tractograms/streamline_operations.py:410]: Remove spaces

**Review by EmmaRenauld — APPROVED** [review]

**Pull request merged**
